File: Utilities/DatabaseUpdater/DatabaseProjectScriptSource.vb
Imports System.Collections.Generic
Public Class DatabaseProjectScriptSource
	Inherits ScriptSource
	Friend path As String
	Friend current As Integer
	Friend fileContent As String()
	Friend currentPath As String
	Friend depth As Integer = 0


	Sub New(ByVal path As String)
		Me.path = path
	End Sub

	Public Overrides Sub FillTables()
		currentPath = New IO.FileInfo(path).Directory.FullName + "\"
		fileContent = System.IO.File.ReadAllLines(path)
		current = 0
		While current < fileContent.Length
			If IsStartOfFolderDefinition() Then
				Select Case GetBitInQuotes()
					Case "Change Scripts"
						AddScriptsFromDirectory(ScriptType.UpdateScript)
					Case "Create Scripts"
						AddCreateScripts()
					Case "Data Scripts"
						AddScriptsFromDirectory(ScriptType.TableData)
				End Select
			End If
			MoveToNextLine()
		End While
	End Sub


	Sub AddCreateScripts()
		MoveToNextLine()
		While Not IsEndOfFolderDefinition()
			If IsStartOfFolderDefinition() Then
				Select Case GetBitInQuotes()
					Case "prc"
						AddScriptsFromDirectory(ScriptType.StoredProcedure)
					Case "viw"
						AddScriptsFromDirectory(ScriptType.View)
					Case "udf"
						AddScriptsFromDirectory(ScriptType.UserDefinedFunction)
					Case "trg"
						AddScriptsFromDirectory(ScriptType.Trigger)
				End Select
			Else
				Throw New Exception("Should only be folder definitions in Create Script folder. Found : " + Line)
			End If
		End While
	End Sub
	Sub AddScriptsFromDirectory(ByVal type As String)
		MoveToNextLine()
		While Not IsEndOfFolderDefinition()
			If IsStartOfFolderDefinition() Then
				AddScriptsFromDirectory(type)
			ElseIf Me.IsScriptDefinition Then
				AddScript(type, currentPath + GetBitInQuotes())
			Else
				Throw New Exception("Should only be script or folder definitions in " + currentPath + ". Found : " + Line)
			End If
			MoveToNextLine()
		End While
	End Sub
	
	Sub AddFolderPathToCurrentPath()
		Me.currentPath += GetBitInQuotes() + "\"
		Me.depth += 1
	End Sub
	Sub MoveToNextLine()
		If Me.IsStartOfFolderDefinition Then AddFolderPathToCurrentPath()
		current += 1
		If Me.IsEndOfFolderDefinition Then GoUpALevel()
	End Sub
	Sub GoUpALevel()
		If Me.depth >= 0 Then
			Me.currentPath = Me.currentPath.Substring(0, Me.currentPath.Length - 1)	'remove trailing \
			Me.currentPath = Me.currentPath.Substring(0, Me.currentPath.LastIndexOf("\") + 1)
			Me.depth -= 1
		End If
	End Sub
	ReadOnly Property Line() As String
		Get
			If current < fileContent.Length Then
				Return Me.fileContent(current).Trim
			Else
				Return ""
			End If
		End Get
	End Property
	Function GetBitInQuotes() As String
		Return Line.Substring(Line.IndexOf("""") + 1, Line.LastIndexOf("""") - Line.IndexOf("""") - 1).Replace("""""", """")
	End Function
	Function IsEndOfFolderDefinition() As Boolean
		Return Line = "End"
	End Function
	Function IsStartOfFolderDefinition() As Boolean
		If Line.Length >= "Begin Folder".Length Then
			Return Line.Substring(0, "Begin Folder".Length) = "Begin Folder"
		Else
			Return False
		End If


	End Function
	Function IsScriptDefinition() As Boolean
		Try
			Return Line.Substring(0, "Script =".Length) = "Script ="
		Catch ex As System.ArgumentOutOfRangeException
			Return False
		End Try

	End Function
End Class

File: Utilities/SqlScriptRunner/ScriptSources/DatabaseProjectScriptSource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NUnit.Framework;
using System.IO;
namespace SqlScriptRunner.ScriptSources
{
    internal class DatabaseProjectScriptSource : ScriptSource
    {

        internal string path;

        internal int current;

        internal string[] fileContent;

        internal string currentPath;

        internal int depth = 0;

        internal DatabaseProjectScriptSource(string path)
        {
            this.path = path;
            currentPath = new FileInfo(path).Directory.FullName + @"\";
            fileContent = File.ReadAllLines(path);
            current = 0;
            while ((current < fileContent.Length))
            {
                if (IsStartOfFolderDefinition())
                {
                    switch (GetBitInQuotes())
                    {
                        case ScriptSource.UpdateScriptDir:
                            AddScriptsFromDirectory(ScriptType.UpdateScript);
                            break;
                        case ScriptSource.CreateScriptDir:
                            AddCreateScripts();
                            break;
                        case ScriptSource.DataScriptDir:
                            AddScriptsFromDirectory(ScriptType.TableData);
							break;
						case ScriptSource.OnSyncScriptDir :
							AddScriptsFromDirectory(ScriptType.OnSync);
                            break;
                    }
                }
                MoveToNextLine();
            }
        }
        string Line
        {
            get
            {
                if ((current < fileContent.Length))
                {
                    return this.fileContent[current].Trim();
                }
                else
                {
                    return "";
                }
            }
        }

        

        void AddCreateScripts()
        {
            MoveToNextLine();
            while (!IsEndOfFolderDefinition())
            {
                if (IsStartOfFolderDefinition())
                {
                    switch (GetBitInQuotes())
                    {
                        case ScriptSource.ProcedureSubDir :
                            AddScriptsFromDirectory(ScriptType.StoredProcedure);
                            break;
                        case ScriptSource.ViewSubDir :
                            AddScriptsFromDirectory(ScriptType.View);
                            break;
                        case ScriptSource.FunctionSubDir :
                            AddScriptsFromDirectory(ScriptType.UserDefinedFunction);
                            break;
                        case ScriptSource.TriggerSubDir:
                            AddScriptsFromDirectory(ScriptType.Trigger);
                            break;
                    }
                }
                else
                {
                    throw new Exception(("Should only be folder definitions in Create Script folder. Found : " + Line));
                }
            }
        }

        void AddScriptsFromDirectory(string type)
        {
            MoveToNextLine();
            while (!IsEndOfFolderDefinition())
            {
                if (IsStartOfFolderDefinition())
                {
                    AddScriptsFromDirectory(type);
                }
                else if (this.IsScriptDefinition())
                {
                    AddScript(type, (currentPath + GetBitInQuotes()));
                }
                else
                {
                    throw new Exception(("Should only be script or folder definitions in "
                                    + (currentPath + (". Found : " + Line))));
                }
                MoveToNextLine();
            }
        }

        void AddFolderPathToCurrentPath()
        {
            this.currentPath = (this.currentPath
                        + (GetBitInQuotes() + "\\"));
            this.depth++;
        }

        void MoveToNextLine()
        {
            if (this.IsStartOfFolderDefinition())
            {
                AddFolderPathToCurrentPath();
            }
            current++;
            if (this.IsEndOfFolderDefinition())
            {
                GoUpALevel();
            }
        }

        internal void GoUpALevel()
        {
            if ((this.depth >= 0))
            {
                this.currentPath = this.currentPath.Substring(0, (this.currentPath.Length - 1));
                // remove trailing \
                this.currentPath = this.currentPath.Substring(0, (this.currentPath.LastIndexOf("\\") + 1));
                this.depth--;
            }
        }

        string GetBitInQuotes()
        {
            return Line.Substring((Line.IndexOf("\"") + 1), (Line.LastIndexOf("\"")
                            - (Line.IndexOf("\"") - 1))).Replace("\"\"", "\"");
        }

        bool IsEndOfFolderDefinition()
        {
            return Line == "End";
        }

        bool IsStartOfFolderDefinition()
        {
        if (Line.Length >= "Begin Folder".Length ){
            return Line.Substring(0, "Begin Folder".Length) == "Begin Folder";
		}else{
			return false;
		}
        }

        bool IsScriptDefinition()
        {
            try
            {
                return Line.Substring(0, "Script =".Length) == "Script =";
            }
            catch (System.ArgumentOutOfRangeException ex)
            {
                return false;
            }
        }
    }

}
